---
title: "RNA-Seq Analysis"
author: "Hector Kroes"
date: "2022-10-12"
output:
  html_document:
    theme: paper
    fig_width: 8
    toc: true
    toc_float: true
    number_sections: true
    link-citations: true
    code_folding: hide
bibliography: bibliography.bib
link-citations: true
csl: citation_style.csl

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# The Dataset

## Characterization

The dataset chosen for the analysis is the following:

- Title: Characterizing gene expression in lung tissue of COPD subjects using RNA-seq
- Overall design: Examination of lung tissue in COPD patients versus normal control
- SRA: SRP041538
- BioProject: PRJNA245811
- GEO: GSE57148

They analyzed gene expression profiling of lung tissue to define molecular pathway of COPD using recent RNA sequencing technology. Lung tissue was obtained from 98 COPD subjects and 91 subjects with normal spirometry. RNA isolated from these samples was processed with RNA-seq using HiSeq 2000. Gene expression measurements were calculated using Cufflinks software.

## The Disease

Chronic obstructive pulmonary disease (COPD) is the third leading cause of death worldwide, causing 3.23 million deaths in 2019 [@noauthor_chronic_2020]. It is a common, preventable and treatable chronic lung disease which affects approximately 300 million men and women worldwide, mainly in low to middle-income countries [@ruvuna_epidemiology_2020].

As it develops, abnormalities in the small airways of the lungs lead to limitation of airflow in and out of the lungs. Several processes cause the airways to become narrow. There may be destruction of parts of the lung, mucus blocking the airways, and inflammation and swelling of the airway lining. It usually manifests itself through cough, wheezing and difficulty breathing.


## Initial data handling

The files were downloaded from recount2 using the study acession id "SRP041538". The data was then scaled and the patients were divided into conditions reflecting whether or not they had COPD, with the reference being the individuals with normal spirometry (NOR). The samples are formatted according to this specific design and then filtered by average gene counts following a threshold adequate for the analysis pipeline used [@sha_effect_2015].

```{r, message=FALSE, warning=FALSE}
# Importing libraries
library(DiagrammeR)
library(recount)
library(DESeq2)
library(ggplot2)
library(tidyverse)
library(ggrepel)
library(DOSE)
library(DT)
require(biomaRt)
library(pheatmap)
library(fgsea)
library(reactome.db)

# Defining study acession
study_acession <- "SRP041538"

# Downloading data, if not already available
if (!file.exists(file.path(study_acession, "rse_gene.Rdata"))) {
  download_study(study_acession)
}

load(file.path(study_acession, "rse_gene.Rdata"))

# Reading and labeling data
rse <- scale_counts(rse_gene)
rse$condition <- sapply(strsplit(as.character(rse$title), "-"), `[`, 2)

# Formatting dataset according to the specific design
dds <- DESeqDataSet(rse, ~condition)
dds$condition <- relevel(dds$condition, ref = "NOR")

# Filtering samples
keepers <- (rowSums(counts(dds)) / ncol(dds)) >= quantile((rowSums(counts(dds)) / ncol(dds)), .2)
dds <- dds[keepers, ]
```

# RNA-Seq Analysis

## Differential expression analysis

We'll perform the differential expression analysis using DESeq2. The referencial group is that of the individuals with normal spirometry (NOR) and the alpha error cuttoff shall be 0.05. Gene symbols are the way they'll be referred to in the next plots and are HGNC symbols whenever possible. The genes exhibited in this following table are those statistically significant and with an absolute logarithmic fold change > 1:

```{r, message=FALSE, warning=FALSE}
dds <- DESeq(dds)

res <- as_tibble(results(dds, alpha = 0.05), rownames = "gene") %>%
  rename(logFC = log2FoldChange) %>%
  rename(FDR = padj)

ensembl <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = "www.ensembl.org")

genes <- gsub("\\..*", "", res$gene)

genemap <- getBM(
  attributes = c("ensembl_gene_id", "hgnc_symbol", "entrezgene_id"),
  filters = "ensembl_gene_id",
  values = genes,
  mart = ensembl
)

check_local_symbol <- function(ref, local_symbols) {
  if (ref %in% local_symbols$group_name) {
    hgnc_symbol <- local_symbols$value[local_symbols$group_name == ref][1]
    if (!hgnc_symbol == "") {
      return(hgnc_symbol)
    } else {
      return(NA)
    }
  } else {
    return(NA)
  }
}

local_symbols <- data.frame(rowData(dds)$symbol)
local_symbols$value[is.na(local_symbols$value)] <- ""
gene_id_list <- NULL

gene_id_lister <- function(ref, genemap) {
  if (ref %in% genemap$ensembl_gene_id) {
    hgnc_symbol <- genemap$hgnc_symbol[genemap$ensembl_gene_id == ref][1]
    if (!hgnc_symbol == "") {
      return(hgnc_symbol)
    } else {
      ref <- check_local_symbol(ref, local_symbols)
      return(ref)
    }
  } else {
    ref <- check_local_symbol(ref, local_symbols)
    return(ref)
  }
}

gene_id_list <- sapply(genes, gene_id_lister, genemap)

res <- res %>%
  add_column(symbol = gene_id_list) %>%
  relocate(symbol, .after = gene)

deseq_genes <- res %>%
  as.data.frame() %>%
  filter(logFC >= log2(2) | logFC <= -log2(2)) %>%
  filter(FDR < 0.05) %>%
  dplyr::select(EnsemblID = gene, symbol, logFC, pvalue, FDR) %>%
  arrange(FDR)
deseq_genes %>%
  datatable()
```

# Data Visualization

## PCA plot

```{r, message=FALSE, warning=FALSE}
res <- res %>%
  mutate(
    Expression = case_when(
      logFC >= log2(2) & FDR <= 0.05 ~ "Up-regulated",
      logFC <= -log2(2) & FDR <= 0.05 ~ "Down-regulated",
      TRUE ~ "Unchanged"
    )
  ) %>%
  mutate(
    Significance = case_when(
      abs(logFC) >= log2(2) & FDR <= 0.05 & FDR > 0.01 ~ "FDR 0.05",
      abs(logFC) >= log2(2) & FDR <= 0.01 & FDR > 0.001 ~ "FDR 0.01",
      abs(logFC) >= log2(2) & FDR <= 0.001 ~ "FDR 0.001",
      TRUE ~ "Unchanged"
    )
  )

tdds <- vst(dds, blind = T)

diffexpgen <- res %>%
  filter(Expression == "Up-regulated" | Expression == "Down-regulated")

tdds <- tdds[rownames(tdds) %in% diffexpgen$gene, ]

pcaData <- plotPCA(tdds, intgroup = "condition", returnData = TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, color = group)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA plot of COPD patients x Healthy individuals")
```

## Volcano plot

```{r, message=FALSE, warning=FALSE}
top_bottom <- function(top, res) {
  top_genes <- bind_rows(
    res %>%
      filter(Expression == "Up-regulated") %>%
      arrange(FDR, desc(abs(logFC))) %>%
      head(top),
    res %>%
      filter(Expression == "Down-regulated") %>%
      arrange(FDR, desc(abs(logFC))) %>%
      head(top)
  )
  return(top_genes)
}

volcano_plot <- ggplot(res, aes(logFC, -log(FDR, 10))) +
  geom_point(aes(color = Significance), size = 2 / 5) +
  xlab(expression("log"[2] * "(Fold change)")) +
  ylab(expression("-log"[10] * "FDR")) +
  scale_color_viridis_d(option = "B") +
  guides(colour = guide_legend(override.aes = list(size = 1.5))) +
  geom_vline(xintercept = c(-1, 1), col = "red") +
  geom_hline(yintercept = -log10(0.05), col = "red") +
  ggtitle("Lung tissue transcriptome of COPD patients x Healthy individuals") +
  geom_label_repel(
    data = top_bottom(40, res),
    mapping = aes(logFC, -log(FDR, 10), label = symbol),
    size = 2,
    max.overlaps = Inf
  )

volcano_plot
```

## Heatmap

```{r, message=FALSE, warning=FALSE}
tddsa <- assay(tdds)
rownames(tddsa) <- gsub("\\..*", "", row.names(tddsa))

filtered_res <- res %>%
  filter(!is.na(symbol))

heat_genes <- top_bottom(10, filtered_res)
heat_genes$gene <- gsub("\\..*", "", heat_genes$gene)

topDE <- tddsa[heat_genes$gene, ]
rownames(topDE) <- heat_genes$symbol

col_annotation <- as.data.frame(colData(tdds)[, "condition"])
colnames(col_annotation) <- "Condition"
rownames(col_annotation) <- colnames(topDE)

row_annotation <- as.data.frame(heat_genes$Expression)
colnames(row_annotation) <- "Expression"
rownames(row_annotation) <- heat_genes$symbol

pheatmap(topDE,
  scale = "row",
  clustering_distance_rows = "correlation",
  annotation_col = col_annotation,
  annotation_row = row_annotation,
  show_colnames = FALSE,
  annotation_names_row = FALSE,
  annotation_names_col = FALSE,
  main = "20 genes with the strongest differential expression"
)
```

## GSEA

```{r, message=FALSE, warning=FALSE}
genemap$entrezgene_id[is.null(genemap$entrezgene_id)] <- NA

entrez_id_lister <- function(ref, genemap) {
  if (ref %in% genemap$ensembl_gene_id) {
    entrez_id <- genemap$entrezgene_id[genemap$ensembl_gene_id == ref][1]
    if (!is.na(entrez_id)) {
      return(entrez_id)
    } else {
      ref <- check_local_symbol(ref, local_symbols)
      return(NA)
    }
  } else {
    ref <- check_local_symbol(ref, local_symbols)
    return(NA)
  }
}

entrez_id_list <- sapply(genes, entrez_id_lister, genemap)

res$entrez_id <- as.character(entrez_id_list)

gsea_data <- subset(res, !is.na(stat)) %>%
  subset(!is.na(FDR))

pathways <- reactomePathways(gsea_data$entrez_id)

ranked_genes <- gsea_data$stat
names(ranked_genes) <- gsea_data$entrez_id

fgseaRes <- fgsea(pathways, ranked_genes, maxSize = 500)

topPathwaysUp <- fgseaRes[NES > 0][head(order(padj), n = 10), pathway]
topPathwaysDown <- fgseaRes[NES < 0][head(order(padj), n = 10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(pathways[topPathways], ranked_genes, fgseaRes,
  gseaParam = 0.5
)
```

# Biological relevance

COPD has two main components in its pathophysiological presentation: emphysema and chronic bronchitis. They are shown in the following flowchart:

```{r, message=FALSE, warning=FALSE}

inf_genes_symb <- c("CXCR1", "CXCR2", "CXCL12", "IL1B", "RELA", "EGFR")
inf_genes_eid <- c("ENSG00000163464.7", "ENSG00000180871.7", "ENSG00000107562.16", "ENSG00000125538.11", "ENSG00000173039.18", "ENSG00000146648.16")
cit <- inf_genes_symb[3:6]

mmps <- (res %>% filter(FDR <= 0.05) %>% filter(logFC > 0) %>% filter(grepl("^MMP", symbol)))$symbol

mucs <- (res %>% filter(FDR <= 0.05) %>% filter(logFC > 1) %>% filter(grepl("MUC", symbol)))$symbol

serpins <- (res %>% filter(!Significance == "Unchanged") %>% filter(grepl("SERPIN", symbol)))
serp_genes_symb <- serpins$symbol
serp_genes_eid <- serpins$gene

DiagrammeR::grViz("digraph flowchart {
      # node definitions with substituted label text
      node [fontname = Helvetica, shape = rectangle]
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']
      tab6 [label = '@@6']
      tab7 [label = '@@7']
      tab8 [label = '@@8']
      tab9 [label = '@@9']
      tab10 [label = '@@10']
      tab11 [label = '@@11']
      tab12 [label = '@@12', style= 'filled', fillcolor = '#E9967A']
      tab13 [label = '@@13', style= 'filled', fillcolor = '#E9967A']
      tab14 [label = '@@14']
      tab15 [label = '@@15']

      # edge definitions with the node IDs
      tab15 -> tab1 -> {tab2 tab4 tab6};
      tab2 -> {tab3 tab4};
      tab3 -> {tab8 tab9 tab10}
      {tab8 tab9 tab10} -> tab12;
      tab4 -> {tab5 tab7} -> tab11;
      tab6 -> tab11;
      tab4 -> tab3;
      tab11-> tab13;
      tab14 -> tab2;
      }

      [1]: 'Inflammation'
      [2]: 'Proteinase–antiproteinase imbalance'
      [3]: 'Destruction of alveolar walls and capillaries'
      [4]: 'Oxidative stress and inflamatory mediators'
      [5]: 'Fibrosis and thickening of bronchiolar walls'
      [6]: 'Mucus hypersecretion and ciliary dysfunction'
      [7]: 'Edema and smooth muscle contration in the airways'
      [8]: 'Enlarged air spaces'
      [9]: 'Impaired gas diffusion'
      [10]: 'Air trapping on expiration'
      [11]: 'Narrowing of small vessels'
      [12]: 'Emphysema'
      [13]: 'Chronic bronchitis'
      [14]: 'Serpin genetic deficiencies'
      [15]: 'Environmental expositions'
      ")
```

I'll present the specific findings of the analysis in terms of how they relate to the course of the development of the disease.

## Emphysema

Environmental exposures such as smoking, secondhand smoke exposure, air pollution and occupational exposure to other dusts, fumes and chemicals are the leading risk factors for the development of COPD [@ruvuna_epidemiology_2020]. They irritate the lungs, leading to an inflammation driven mainly by neutrophils and lymphocytes. 

When recruited to the bronchioli and alveoli, neutrophils release elastases, enzymes that degrade the elastin in the alveolar walls and capillaries of the lungs. Neutrophils also produce reactive oxygen species (ROS), inflammatory mediators and cytokines that promote metalloproteinase (MMPs) activity which is associated with the destruction of the extracellular matrix (ECM) components and the aberrant remodelling of damaged alveoli while inhibiting tissue inhibitors of metalloproteinases (TIMPs) activity which is associated to the maintenance of ECM integrity. This constitutes the classic protease-antiprotease imbalance that leads to emphysema by the destruction of alveolar walls and capillaries with an enlargement of air spaces, impaired gas diffusion and air trapping on expiration [@laurell_electrophoretic_2013].

The neutrophil pathologic role in COPD highly depends on the activity of the CXCL-8-CXCR1/2 axis [@ha_role_2017], which was found to be upregulated in patients with the disease. Other genes in the same axis, such as `r cit` are also upregulated. As for the metalloproteinases, we found that several of them (`r mmps`) were upregulated. No TIMP was found to be downregulated.

```{r, message=FALSE, warning=FALSE}
relexp <- function(eid, res) {
  index <- res %>%
    filter(gene == eid) %>%
    dplyr::select(logFC)
  return(2**index$logFC)
}

getpvalue <- function(symb, res) {
  padj <- res %>%
    filter(symbol == symb) %>%
    dplyr::select(FDR)
  padj <- round(padj$FDR, digits = 3)
  if (padj == 0) {
    return("p-value\n<0.001")
  } else {
    return(paste0("p-value\n", padj))
  }
}

c1 <- c(rep(inf_genes_symb, 2))
group <- c(c(rep("NOR", length(inf_genes_symb))), c(rep("COPD", length(inf_genes_symb))))
c3 <- c(c(rep(1, length(inf_genes_symb))), c(sapply(inf_genes_eid, relexp, res)))
c4 <- sapply(inf_genes_symb, getpvalue, res)

annot1 <-
  tibble(
    variable = inf_genes_symb,
    label = c4,
    xpos = c(1:length(inf_genes_symb)),
    ypos = Inf,
    vjustvar = 2
  )

df <- data.frame(c1, group, c3)
df$group <- factor(df$group, levels = c("NOR", "COPD"))

bargraph <- ggplot(data = df) +
  geom_bar(aes(
    x = c1,
    y = c3,
    fill = group,
    color = group
  ),
  stat = "identity",
  position = position_dodge()
  ) +
  geom_text(data = annot1, aes(label = label, x = xpos, y = ypos, vjust = vjustvar)) +
  labs(
    x = "Genes", y = "Relative expression",
    title = "Relative expression of genes in CXCL-8-CXCR1/2 axis"
  ) +
  expand_limits(x = c(0, 0), y = c(0, 2.5))
bargraph
```

## Chronic bronchitis

The second of the main pillars for the presentation of COPD is chronic bronchitis. The whole inflammatory environment at the lungs leads to mucus hypersecretion, ciliary dysfunction, fibrosis, thickening of the bronchiolar walls, oedema and smooth muscle contraction in the airways, producing a narrowing of the small airways that characterizes COPD’s chronic bronchitis.

We can observe the increase in mucin expression in COPD patients. From the nine main human mucin genes, we find that three of them (`r mucs`) are upregulated. MUC5AC is the main gene previously reported overexpressed in COPD patients’ sputum [@chadwick_mucus_2008]. 

## Mitochondrial dysfunction

A less well-known dimension of the pathophysiology of COPD is its association with mitochondrial dysfunction. When faced with oxidative stress, the lung mitochondria present a decrease in membrane potential and abnormal patterns of expression regulation [@wiegman_oxidative_2015]. 

The GEAS analysis also demonstrates that many mitochondria-related pathways are altered in COPD patients. Prohibitin (PHB-1) has an important role in mtDNA regulation by stabilising TFAM protein, a known protein component of the mitochondrial nucleoids [@kasashima_human_2008]. In this analysis, PHB-1 was found to be down-regulated and TFAM up-regulated in COPD patients, a combination that can explain at least partially the mitochondrial dysfunction. This increases the inflammatory response even more.

```{r, message=FALSE, warning=FALSE}
mito_plot <- plotEnrichment(pathways[["Mitochondrial translation"]], ranked_genes) + labs(title = "Mitochondrial translation enrichment")

mito_plot
```

## Serpins

Serpins are a superfamily of serine protease inhibitors whose genetic deficiency can cause COPD. It usually presents differently from other causes of COPD, as it tends to promote the formation of panacinar emphysema in contrast to centrilobular and paraseptal emphysema of other aetiologies Perhaps the best-known of these deficiencies is alpha-1 antitrypsin (SERPINA1) deficiency. 

In our dataset, there’s no distinction for each patient’s COPD aetiology. But after this analysis, we can verify that many representatives of the serpin superfamily are shown to be upregulated in COPD patients (`r serp_genes_symb`), which is the normal physiological reaction to inflammation [@kelly-robinson_serpin_2021], showing that probably our COPD group is mostly represented by individuals without serpin deficiencies. 

```{r, message=FALSE, warning=FALSE}
c1 <- c(rep(serp_genes_symb, 2))
group <- c(c(rep("NOR", nrow(serpins))), c(rep("COPD", nrow(serpins))))
c3 <- c(c(rep(1, nrow(serpins))), c(sapply(serp_genes_eid, relexp, res)))
c4 <- sapply(serp_genes_symb, getpvalue, res)

annot <-
  tibble(
    variable = serp_genes_symb,
    label = c4,
    xpos = c(1:nrow(serpins)),
    ypos = Inf,
    vjustvar = 2
  )

df <- data.frame(c1, group, c3)
df$group <- factor(df$group, levels = c("NOR", "COPD"))

bargraph <- ggplot(data = df) +
  geom_bar(aes(
    x = c1,
    y = c3,
    fill = group,
    color = group
  ),
  stat = "identity",
  position = position_dodge()
  ) +
  geom_text(data = annot, aes(label = label, x = xpos, y = ypos, vjust = vjustvar)) +
  labs(
    x = "Genes", y = "Relative expression",
    title = "Relative expression of differentially expressed serpins"
  ) +
  expand_limits(x = c(0, 0), y = c(0, 3))
bargraph
```

Comparative expression profiles of COPD patients of different aetiologies could be valuable in understanding better the disease and designing specific treatments according to their specificities. 

# Conclusion

We can clearly see that there are marked differences in the lung tissue expression profile of patients with and without COPD. And our findings were very consistent with the data already known about the disease and further strengthens the current theories about its pathophysiology. Future studies could focus on the differences between COPD's distinct aetiologies leading to more personalized approaches to these patients' care.


# Bibliography 

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
