---
title: "RNA-Seq Analysis"
author: "Hector Kroes"
date: "2022-10-12"
output:
  html_document:
    theme: paper
    fig_width: 8
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
bibliography: bibliography.bib

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# The Dataset

## Characterization

The dataset chosen for the analysis is the following:

- Title: Characterizing gene expression in lung tissue of COPD subjects using RNA-seq
- Overall design: Examination of lung tissue in COPD patients versus normal control
- SRA: SRP041538
- BioProject: PRJNA245811
- GEO: GSE57148

They analyzed gene expression profiling of lung tissue to define molecular pathway of COPD using recent RNA sequencing technology. Lung tissue was obtained from 98 COPD subjects and 91 subjects with normal spirometry. RNA isolated from these samples was processed with RNA-seq using HiSeq 2000. Gene expression measurements were calculated using Cufflinks software.

## The Disease

Chronic obstructive pulmonary disease (COPD) is the third leading cause of death worldwide, causing 3.23 million deaths in 2019. It is a common, preventable and treatable chronic lung disease which affects approximately 300 million men and women worldwide, mainly in low to middle-income countries.

As it develops, abnormalities in the small airways of the lungs lead to limitation of airflow in and out of the lungs. Several processes cause the airways to become narrow. There may be destruction of parts of the lung, mucus blocking the airways, and inflammation and swelling of the airway lining. It usually manifests itself through cough, wheezing and difficulty breathing.

## Initial data handling

The files were downloaded from recount2 using the study acession id "SRP041538". The data was then scaled and the patients were divided into conditions reflecting whether or not they had COPD, with the reference being the individuals with normal spirometry (NOR). The samples are formatted according to this specific design and then filtered by average gene counts following a threshold adequate for the analysis pipeline used [@sha_effect_2015].

```{r, message=FALSE, warning=FALSE}
# Importing libraries
library(recount)
library(DESeq2)
library(ggplot2)
library(tidyverse)
library(ggrepel)
library(DOSE)
library(DT)
require(biomaRt)
library(pheatmap)
library(fgsea)
library(reactome.db)

# Defining study acession
study_acession <- "SRP041538"

# Downloading data, if not already available
if (!file.exists(file.path(study_acession, "rse_gene.Rdata"))) {
  download_study(study_acession)
}

load(file.path(study_acession, "rse_gene.Rdata"))

# Reading and labeling data
rse <- scale_counts(rse_gene)
rse$condition <- sapply(strsplit(as.character(rse$title), "-"), `[`, 2)

# Formatting dataset according to the specific design
dds <- DESeqDataSet(rse, ~condition)

# Filtering samples
keep <- (rowSums(counts(dds)) / ncol(dds)) >= quantile((rowSums(counts(dds)) / ncol(dds)), .2)
dds <- dds[keep, ]

dds$condition <- relevel(dds$condition, ref = "NOR")
```

# RNA-Seq Analysis

## Differential expression analysis

We'll perform the differential expression analysis using DESeq2. The referencial group is that of the individuals with normal spirometry (NOR) and the alpha error cuttoff shall be 0.05. Gene symbols are the way they'll be referred to in the next plots and are HGNC symbols whenever possible (in the absence of an equivalent HGNC symbol, the standard EnsemblID shall be used). The genes exhibited in this following table are those statiscally significant and with an absolute fold change > 2:

```{r, message=FALSE, warning=FALSE}

rse$condition <- sapply(strsplit(as.character(rse$title), "-"), `[`, 2)

dds <- DESeqDataSet(rse, ~condition)

dds$treatment <- relevel(dds$condition, ref = "NOR")
dds <- DESeq(dds)

res <- as_tibble(results(dds, alpha = 0.05), rownames = "gene") %>%
  rename(logFC = log2FoldChange) %>%
  rename(FDR = padj)

ensembl <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = "www.ensembl.org")

genes <- gsub("\\..*", "", res$gene)

genemap <- getBM(
  attributes = c("ensembl_gene_id", "hgnc_symbol", "entrezgene_id"),
  filters = "ensembl_gene_id",
  values = genes,
  mart = ensembl
)

check_local_symbol <- function(ref, local_symbols) {
  if (ref %in% local_symbols$group_name) {
    hgnc_symbol <- local_symbols$value[local_symbols$group_name == ref][1]
    if (!hgnc_symbol == "") {
      return(hgnc_symbol)
    } else {
      return(NA)
    }
  } else {
    return(NA)
  }
}

local_symbols <- data.frame(rowData(dds)$symbol)
local_symbols$value[is.na(local_symbols$value)] <- ""
gene_id_list <- NULL


for (ref in genes) {
  if (ref %in% genemap$ensembl_gene_id) {
    hgnc_symbol <- genemap$hgnc_symbol[genemap$ensembl_gene_id == ref][1]
    if (!hgnc_symbol == "") {
      gene_id_list <- append(gene_id_list, hgnc_symbol)
    } else {
      ref <- check_local_symbol(ref, local_symbols)
      gene_id_list <- append(gene_id_list, ref)
    }
  } else {
    ref <- check_local_symbol(ref, local_symbols)
    gene_id_list <- append(gene_id_list, ref)
  }
}

res <- res %>%
  add_column(symbol = gene_id_list) %>%
  relocate(symbol, .after = gene)

deseq_genes <- res %>%
  as.data.frame() %>%
  filter(FDR < 0.05) %>%
  filter(logFC >= log2(2) | logFC <= -log2(2)) %>%
  dplyr::select(EnsemblID = gene, symbol, logFC, pvalue, FDR) %>%
  arrange(FDR)
deseq_genes_count <- nrow(deseq_genes)
output_note <- paste0("\nTotal differential expressed genes with adjusted p<0.05: ", deseq_genes_count, "\n(top 1000 shown above)")
deseq_genes %>%
  datatable()
```

# Data Visualization

## PCA plot

```{r, message=FALSE, warning=FALSE}
tdds <- vst(dds, blind = T)

pcaData <- plotPCA(tdds, intgroup = "condition", returnData = TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot(pcaData, aes(PC1, PC2, color = group)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA plot of COPD patients x Healthy individuals")
```

## Volcano plot

```{r, message=FALSE, warning=FALSE}
res <- res %>%
  mutate(
    Expression = case_when(
      logFC >= log2(2) & FDR <= 0.05 ~ "Up-regulated",
      logFC <= -log2(2) & FDR <= 0.05 ~ "Down-regulated",
      TRUE ~ "Unchanged"
    )
  ) %>%
  mutate(
    Significance = case_when(
      abs(logFC) >= log2(2) & FDR <= 0.05 & FDR > 0.01 ~ "FDR 0.05",
      abs(logFC) >= log2(2) & FDR <= 0.01 & FDR > 0.001 ~ "FDR 0.01",
      abs(logFC) >= log2(2) & FDR <= 0.001 ~ "FDR 0.001",
      TRUE ~ "Unchanged"
    )
  )

top_bottom <- function(top, res) {
  top_genes <- bind_rows(
    res %>%
      filter(Expression == "Up-regulated") %>%
      arrange(FDR, desc(abs(logFC))) %>%
      head(top),
    res %>%
      filter(Expression == "Down-regulated") %>%
      arrange(FDR, desc(abs(logFC))) %>%
      head(top)
  )
  return(top_genes)
}

volcano_plot <- ggplot(res, aes(logFC, -log(FDR, 10))) +
  geom_point(aes(color = Significance), size = 2 / 5) +
  xlab(expression("log"[2] * "(Fold change)")) +
  ylab(expression("-log"[10] * "FDR")) +
  scale_color_viridis_d(option = "B") +
  guides(colour = guide_legend(override.aes = list(size = 1.5))) +
  geom_vline(xintercept = c(-1, 1), col = "red") +
  geom_hline(yintercept = -log10(0.05), col = "red") +
  ggtitle("Lung tissue transcriptome of COPD patients x Healthy individuals") +
  geom_label_repel(
    data = top_bottom(40, res),
    mapping = aes(logFC, -log(FDR, 10), label = symbol),
    size = 2,
    max.overlaps = Inf
  )

volcano_plot
```

## Heatmap

```{r, message=FALSE, warning=FALSE}
tddsa <- assay(tdds)
rownames(tddsa) <- gsub("\\..*", "", row.names(tddsa))

filtered_res <- res %>%
  filter(!is.na(symbol))

heat_genes <- top_bottom(10, filtered_res)
heat_genes$gene <- gsub("\\..*", "", heat_genes$gene)

topDE <- tddsa[heat_genes$gene, ]
rownames(topDE) <- heat_genes$symbol

col_annotation <- as.data.frame(colData(tdds)[, "condition"])
colnames(col_annotation) <- "Condition"
rownames(col_annotation) <- colnames(topDE)

row_annotation <- as.data.frame(heat_genes$Expression)
colnames(row_annotation) <- "Expression"
rownames(row_annotation) <- heat_genes$symbol

pheatmap(topDE,
  scale = "row",
  clustering_distance_rows = "correlation",
  annotation_col = col_annotation,
  annotation_row = row_annotation,
  show_colnames = FALSE,
  annotation_names_row = FALSE,
  annotation_names_col = FALSE,
  main = "20 genes with the strongest differential expression"
)
```

## GSEA

```{r, message=FALSE, warning=FALSE}
genemap$entrezgene_id[is.null(genemap$entrezgene_id)] <- NA

entrez_id_list <- NULL

for (ref in genes) {
  if (ref %in% genemap$ensembl_gene_id) {
    entrez_id <- genemap$entrezgene_id[genemap$ensembl_gene_id == ref][1]
    if (!is.na(entrez_id)) {
      entrez_id_list <- append(entrez_id_list, entrez_id)
    } else {
      entrez_id_list <- append(entrez_id_list, NA)
    }
  } else {
    entrez_id_list <- append(entrez_id_list, NA)
  }
}

res$entrez_id <- as.character(entrez_id_list)

gsea_data <- subset(res, !is.na(stat)) %>%
  subset(!is.na(FDR))

pathways <- reactomePathways(gsea_data$entrez_id)

ranked_genes <- gsea_data$stat
names(ranked_genes) <- gsea_data$entrez_id

fgseaRes <- fgsea(pathways, ranked_genes, maxSize = 500)

topPathwaysUp <- fgseaRes[NES > 0][head(order(padj), n = 10), pathway]
topPathwaysDown <- fgseaRes[NES < 0][head(order(padj), n = 10), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(pathways[topPathways], ranked_genes, fgseaRes,
  gseaParam = 0.5
)
```

# Biological Relevance

grViz("digraph flowchart {
      # node definitions with substituted label text
      node [fontname = Helvetica, shape = rectangle]        
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']
      tab6 [label = '@@6']
      tab7 [label = '@@7']
      tab8 [label = '@@8']
      tab9 [label = '@@9']
      tab10 [label = '@@10']
      tab11 [label = '@@11']
      tab12 [label = '@@12', fillcolor = red]
      tab13 [label = '@@13', fillcolor = red]

      # edge definitions with the node IDs
      tab1 -> tab2 -> tab4;
      tab2 -> tab3 -> tab8 -> tab12;
      tab3 -> tab9 -> tab12;
      tab3 -> tab10 -> tab12;
      tab1 -> tab4 -> tab5 -> tab11;
      tab4 -> tab7 -> tab11;
      tab1 -> tab6 -> tab11;
      tab11-> tab13;
      }

      [1]: 'Inflammation'
      [2]: 'Proteinase–antiproteinase imbalance'
      [3]: 'Destruction of alveolar walls and capillaries'
      [4]: 'Oxidative stress and inflamatory mediators'
      [5]: 'Fibrosis and thickening of bronchiolar walls'
      [6]: 'Mucus hypersecretion and ciliary dysfunction'
      [7]: 'Edema and smooth muscle contration in the airways'
      [8]: 'Enlarged air spaces'
      [9]: 'Impaired gas diffusion'
      [10]: 'Air trapping on expiration'
      [11]: 'Narrowing of small vessels'
      [12]: 'Emphysema'
      [13]: 'Chronic bronchitis'
      ")

## Proteinase–antiproteinase imbalance

A proteinase–antiproteinase imbalance has been the most widely accepted theory for the pathogenesis of pulmonary emphysema since the discovery of the α-1-antitrypsin deficiency [@laurell_electrophoretic_2013]. 

In the GEAS plot, we can see the comparatively low enrichment in COPD patients of genes related to the maintenance of the extracellular matrix organization. This represents the smaller expression of antiproteinases and other constitutive elements that are key to the physiological maintenance of the ECM.

Meanwhile, we find that MMP1 (interstitial collagenase) is upregulated in patients with COPD. Matrix metalloproteinases (MMPs) comprise a structurally and functionally related family of proteolytic enzymes, which play an essential in ECM remodelling associated with inflammation. This corroborates its importance in airway inflammation and the development of emphysema already reported in multiple human and animal studies [@joos_role_2002].

## Mitochondrial dysfunction

The GEAS analysis also demonstrates that many mitochondria-related pathways are altered, especially enriched in COPD cases.

Prohibitin (PHB-1) has an important role in mtDNA regulation by stabilizing TFAM protein, a known protein component of the mitochondrial nucleoids [@kasashima_human_2008]. In this analysis, PHB-1 was found to be down-regulated and TFAM up-regulated in COPD patients, a combination that can explain at least partially the mitochondrial dysfunction.

## Inflammation 

Airway inflammation is an important factor in the pathogenesis of COPD as a great part of the changes in the mechanical properties of the lung such as mucosal edema, increased airway secretion and airway remodelling can be attributed to this process.

In this dataset, CXCR1 and CXCR2 were found to be up-regulated, which is compatible with the previous studies that affirmed those chemokine receptors were an important part of the inflammatory response in the bronchial mucosa from patients with COPD during an acute exacerbation. CXCR3 was also up-regulated, probably reflecting the intensified expression of IFN-γ in peripheral airway mucosal cells [@larsson_aspects_2007].

## Susceptibility

Some genes are described in the literature as being risk factors for the development of COPD. One of the main candidates is SERPINA-3, which was also found to be up-regulated in COPD patients in this analysis.

# Bibliography 

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
